### PR TITLE
Bg schema

### PR DESCRIPTION
**Pull request by rauf543**
# Pull Request
QUICK EDIT: First time doing a PR, so if something is wrong for the format, or with the code, or with the way I have committed etc... please let me know and I will try my best to fix it
Overview

This pull request introduces a new feature to dynamically manage combat background schemas for events and other use cases without requiring hardcoding. It enables admins and event/content staff to add, edit, delete, and activate background schemas through a front-end interface, with changes seamlessly reflected in the combat system.

Back-End Changes

Updated determineArenaBackground and determineCombatBackground:

Replaced constant-based logic with dynamic queries to a new database table: backgroundSchemas.
Functions now query the table to find the active schema and fetch its schema field, which redirects to the appropriate link.
Once fetched, the battle logic remains unchanged.
Database Integration:

Added the backgroundSchemas table to store schema data, including an active status flag.
Supports CRUD operations and status toggling for dynamic updates.
Refactored Function Calls:

Updated all existing calls to determineArenaBackground and determineCombatBackground to align with the new dynamic functionality.
Front-End Changes

New Admin Page:

[QA Testing Checklist.pdf](https://github.com/user-attachments/files/17893294/QA.Testing.Checklist.pdf)

Added a page under the Manuals > Combat section called Background Schemas.
Functionality includes:
Adding new background schemas.
Editing existing schemas.
Deleting unwanted schemas.
Setting active schemas for immediate use in events.
Permission Locking:

Restricted access to this page based on specific permissions.
Permissions required: •	

User: No Access  ✔
Coding Admin: Full CRUD access with ability to change active schema. ✔
Content Admin: Full CRUD access with ability to change active schema. ✔
Moderator Admin: Limited CRUD Access, no ability to change active Schema or delete schemas. ✔
Head Moderator/Moderator: No Access ✔
Content/Event: Limited CRUD Access, no ability to change active Schema or delete schemas. ✔

Summary
This change provides greater flexibility and efficiency for managing combat backgrounds, especially for dynamic events. It simplifies staff workflows by eliminating the need for code-level changes while maintaining secure access control.
## License

By making this pull request, I confirm that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the Studie-Tech ApS organization has the copyright to use and modify my contribution for perpetuity.

### CODE REVIEW

**Comment by MathiasGruber** [comment]
First impression is this looks awesome!.. I will review it as soon as possible - I'm down with a cold these days, so my head is not clear enough right now, but will prioritize this as soon as I can! Once merged, this definitely rewards a good chunk of reputation points - which account would you like these to go to?

File: app/src/validators/backgroundSchema.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { z } from "zod";
+
+export const BackgroundSchemaValidator = z.object({
+  name: z.string().min(1).max(191),
+  description: z.string().min(1).max(191),
+  isActive: z.boolean().default(false),
+  schema: z.object({
```
Comment (MathiasGruber) [inline app/src/validators/backgroundSchema.ts]: Extract `schema` to a separate validator, which is also exported and used here, i.e. 

```
export const bgSchema = z.object({
    ocean: z.string().url(),
    ice: z.string().url(),
    dessert: z.string().url(),
    ground: z.string().url(),
    arena: z.string().url(),
    default: z.string().url(),
  })```
  
  Then we can use this validator & its inferred type in the `schema.ts` file

File: app/drizzle/schema.ts [hunk 2]
```diff
@@ -2260,3 +2260,39 @@ export const userRequestRelations = relations(userRequest, ({ one }) => ({
...
+
+export const backgroundSchema = mysqlTable(
+  "backgroundSchema",
+  {
+    id: varchar("id", { length: 191 })
+      .primaryKey()
+      .notNull()
+      .default(sql`(UUID())`),
+    schema: json("schema")
+      .$type<{
```
Comment (MathiasGruber) [inline app/drizzle/schema.ts]: Reuse the inferred type from the validated file - so that we only need to maintain it one place :)

File: app/pnpm-lock.yaml
Comment (MathiasGruber) [inline app/pnpm-lock.yaml:1]: Why did the lock file change if we don't have new packages?

File: app/src/app/manual/combat/backgroundSchema/edit/[schemaId]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    ocean: schema?.schema?.ocean || "",
+    ice: schema?.schema?.ice || "",
+    dessert: schema?.schema?.dessert || "",
+    ground: schema?.schema?.ground || "",
+    arena: schema?.schema?.arena || "",
+    default: schema?.schema?.default || "",
+  });
+
+  // Define the validation schema using zod
+  const BackgroundSchemaValidator = z.object({
```
Comment (MathiasGruber) [inline app/src/app/manual/combat/backgroundSchema/edit/[schemaId]/page.tsx]: Reuse from the `backgroundSchema.ts` validators file?

File: app/src/app/manual/combat/backgroundSchema/edit/[schemaId]/page.tsx [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  schema?: Schema;
+}
+
+const EditBackgroundSchemaForm: React.FC<EditBackgroundSchemaFormProps> = ({
+  schema,
+  refetch,
+}) => {
+  const router = useRouter();
+  const utils = api.useUtils();
+  const [imageUrls, setImageUrls] = useState<Record<string, string>>({
```
Comment (MathiasGruber) [inline app/src/app/manual/combat/backgroundSchema/edit/[schemaId]/page.tsx]: Maybe also re-use from the validators file, if we add default values of "" to the zod schema? Might not work, not sure.

File: app/src/server/api/routers/backgroundSchema.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      .select()
+      .from(backgroundSchema)
+      .where(like(backgroundSchema.name, "New Background Schema%"))
+      .orderBy(desc(backgroundSchema.createdAt));
+
+    let baseName = "New Background Schema";
+    let newName = baseName;
+    let suffix = 1;
+    const id = nanoid();
+
```
Comment (MathiasGruber) [inline app/src/server/api/routers/backgroundSchema.ts:61]: Permission check missing?

File: app/src/server/api/routers/backgroundSchema.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  }),
+
+  update: protectedProcedure
+    .input(
+      BackgroundSchemaValidator.extend({
+        id: z.string(),
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      const { id, ...data } = input;
```
Comment (MathiasGruber) [inline app/src/server/api/routers/backgroundSchema.ts]: Permission check missing?

File: app/src/server/api/routers/backgroundSchema.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        })
+        .where(eq(backgroundSchema.id, id));
+
+      return { message: "Schema updated successfully", success: true };
+    }),
+
+  delete: protectedProcedure
+    .input(z.object({ id: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      await ctx.drizzle
```
Comment (MathiasGruber) [inline app/src/server/api/routers/backgroundSchema.ts]: Permission check missing?

File: app/src/server/api/routers/backgroundSchema.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        .where(eq(backgroundSchema.id, input.id));
+
+      return { message: "Schema deleted successfully", success: true };
+    }),
+
+  activate: protectedProcedure
+    .input(z.object({ id: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      const { id } = input;
+
```
Comment (MathiasGruber) [inline app/src/server/api/routers/backgroundSchema.ts]: Permission check missing?

File: app/src/server/api/routers/combat.ts [hunk 9]
```diff
@@ -552,6 +543,7 @@ export const combatRouter = createTRPCRouter({
       }
       // Determine battle background
       if (selectedAI) {
+        const background = await determineArenaBackground(ctx.drizzle, "arena");
```
Comment (MathiasGruber) [inline app/src/server/api/routers/combat.ts]: I'm wondering if we can figure out a way in which the queries for fetching background & user information (within initiateBattle) can be run in parallel within Promise.all.. every time we introduce sequential queries, we increase the function call time significantly, and the intiate battle call is one which is quite frequent.

File: das
Comment (MathiasGruber) [inline das:1]: Don't think this should be comitted? :)

**Review by MathiasGruber — DISMISSED** [review]
Overall super good work, and very awesome feature.. I've not tried spinning it up, but I thought all this work deserved a swift review.. I found a few places where I think we could make it better - specifically in trying to have a one source of truth for the schema and validators, and then also performance in terms of running parallel instead of sequential queries during battle initiation

**Comment by rauf543** [comment]
Thank you for the kind words! I am still not too experienced with this technology so appreciate all of the comments. Ill work on the raised issues, most seem pretty straightforward to resolve.

**Comment by rauf543** [comment]
Pushed requested changes, adjusted initiateBattle to work with a parallel promise, added permissions checks for the schema router aswell as tried to use one source of truth (validator) for all schema checks.

**Comment by rauf543** [comment]
> Once merged, this definitely rewards a good chunk of reputation points - which account would you like these to go to?

Thank you for the unexpected but welcome reward! my ign is Dirtyhands.

**Comment by rauf543** [comment]
edge case bug identified.

When you edit a schema, it kicks you back to the getAll page, if you press edit again, it shows you the previous, unedited fields. Probably some other state I need to invalidate. Going to bed right now so will check more on this tommorow.

The bug is minor and after a refresh all is how it should be.

**Comment by MathiasGruber** [comment]
Will review this tonight when I have time! 🤘

**Comment by MathiasGruber** [comment]
I made a mistake and merged this prematurely.. don't worry, I'll fix up the errors and get it working again. Will still upload reputation points to your account once I have it fixed up :)

**Comment by rauf543** [comment]
Not a problem, don’t worry too much about the reps, I do it to learn and for fun, your guidance along the way is much more valuable, thank you for that.

I just wanted to ask, what was the problem? Just so I make sure to commit better code next time so it can be merged straight away.

**Comment by MathiasGruber** [comment]
Just a lot of type errors on the build.. if you run a `make pnpm build` it'll try to build locally, which should reveal all missing type errors :)

**Comment by MathiasGruber** [comment]
Here's the fix I applied to make it compile (and a bit of UX alignment) https://github.com/MathiasGruber/TheNinjaRPG/commit/7c72ab41e75c79e1c2fcff895da1f04041b9e254

Will send 200 reputation points to your account, thanks for the contribution :)

**Pull request merged**